src/coreclr/src/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs

```diff
@@ -3957,112 +3951,181 @@ private void CreateInstanceCheckThis()
-            internal CtorDelegate? _ctor;
+            // The managed calli to the newobj routine, plus its first argument (MethodTable*).
+            private readonly delegate*<MethodTable*, object?> _pfnNewobj;
+            private readonly MethodTable* _pMT;
 
-            // Lazy initialization was performed
-            internal volatile bool _isFullyInitialized;
+            // The managed calli to the parameterless ctor, plus a state object.
+            // State object depends on the stub being called.
+            private readonly delegate*<object?, IntPtr, void> _pfnCtorStub;
+            private readonly IntPtr _ctorStubState;
 
-            private static ConstructorInfo? s_delegateCtorInfo;
+#if DEBUG
+            private readonly RuntimeType _originalRT;
+#endif
 
-            internal ActivatorCache(RuntimeMethodHandleInternal rmh)
+            internal ActivatorCache([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)] RuntimeType rt)
             {
-                _hCtorMethodHandle = rmh;
-            }
+                Debug.Assert(rt != null);
 
-            private void Initialize()
-            {
-                if (!_hCtorMethodHandle.IsNullHandle())
+#if DEBUG
+                _originalRT = rt;
+#endif
+
+                _pfnCtorStub = &CtorNoopStub; // by default, no ctor is run
+                _ctorStubState = IntPtr.Zero;
+
+                _pfnNewobj = RuntimeTypeHandle.GetNewobjHelperFnPtr(rt, out _pMT, unwrapNullable: false);
+
+                RuntimeMethodHandleInternal defaultCtorRMH = RuntimeTypeHandle.GetDefaultConstructor(rt);
+
+                if (_pMT->IsValueType)
                 {
-                    _ctorAttributes = RuntimeMethodHandle.GetAttributes(_hCtorMethodHandle);
+                    if (_pMT->IsNullable)
+                    {
+                        // Activator.CreateInstance returns null given typeof(Nullable<T>).
+
+                        static object? ReturnNull(MethodTable* pMT) => null;
+                        _pfnNewobj = &ReturnNull;
+                        _pMT = null;
+
+                        _pfnCtorStub = &CtorNoopStub;
+                        _ctorStubState = IntPtr.Zero;
+                    }
+                    else if (_pMT->HasDefaultConstructor)
+                    {
+                        // ValueType with explicit parameterless ctor typed as (ref T) -> void.
+                        // We'll pass the actual ctor address as the state object, then create
+                        // an unboxing stub so that we can pass the boxed value to it.
 
-                    // The default ctor path is optimized for reference types only
-                    ConstructorInfo delegateCtorInfo = s_delegateCtorInfo ??= typeof(CtorDelegate).GetConstructor(new Type[] { typeof(object), typeof(IntPtr) })!;
+                        static void ValueTypeUnboxingStub(object? @this, IntPtr state)
```

Comment:
You should not need to do this indirection and `_ctorStubState` if you let runtime create the unboxing function pointer for the ctor (e.g. make it out argument in GetNewObjHelper). It should be something like:

```
MethodDesc::FindOrCreateAssociatedMethodDesc(pMT->GetDefaultConstructor(),
                                             pMT,
                                             TRUE /* no BoxedEntryPointStub */,
                                             Instantiation(), /* no method instantiation */
                                             FALSE /* no allowInstParam */)->GetMultiCallableAddrOfCode();
```

Reply:
Won't this spin up the JIT? I tried to avoid that as much as possible elsewhere throughout this code.

Alternative design: The managed stub function doesn't take a state parameter, but it instead takes the boxed value type and reads the MethodTable\* straight from the boxed object, finds the parameterless ctor, and dispatches to it via `delegate* managed<ref byte, void>`. This means that creating value types with explicit parameterless ctors becomes a more expensive, but: (a) this should be exceedingly rare, (b) no other code paths pay this cost, and (c) we don't need to fire up the JIT.

Reply:
For non-generic structs, it won't spin up the JIT.

For generics structs, it may spin up a JIT to produce the instantiating unboxing stub, but the existing code may spin up the JIT as well to produce instantiating stub. So not much difference there.

Reply:
Actually, it should not spin up a JIT for the generics structs eithers.